examples/example_two_dimensional_peak.py

```diff
@@ -0,0 +1,167 @@
+x = np.random.rand(npoints)*10 - 4
+y = np.random.rand(npoints)*5 - 3
+z = gaussian2d(x, y, amplitude=30, centerx=2, centery=-.5, sigmax=.6, sigmay=.8)
+z += 2*(np.random.rand(*z.shape)-.5)
+error = np.sqrt(z+1)
+###############################################################################
+# To plot this, we can interpolate the data onto a grid.
+X, Y = np.meshgrid(np.linspace(x.min(), x.max(), 100),
+                   np.linspace(y.min(), y.max(), 100))
+Z = griddata((x, y), z, (X, Y), method='linear', fill_value=0)
+
+fig, ax = plt.subplots()
+art = ax.pcolor(X, Y, Z)
+cb = plt.colorbar(art, ax=ax)
+ax.set_xlabel('x')
+ax.set_ylabel('y')
+cb.set_label('z')
+###############################################################################
+# In this case, we can use a built-in model to fit
+model = lmfit.models.Gaussian2dModel()
+params = model.guess(z, x, y)
+result = model.fit(z, x=x, y=y, params=params, weights=1/error)
+result
+###############################################################################
+# To check the fit, we can evaulate the function on the same grid we used
+# before and make plots of the data, the fit and the difference between the two.
+fig, axs = plt.subplots(2, 2, figsize=(10, 10))
+
+vmax = np.nanpercentile(Z, 99.9)
+
+ax = axs[0, 0]
+art = ax.pcolor(X, Y, Z, vmin=0, vmax=vmax)
+cb = plt.colorbar(art, ax=ax, label='z')
+ax.set_title('Data')
+
+ax = axs[0, 1]
+fit = model.func(X, Y, **result.best_values)
+art = ax.pcolor(X, Y, fit, vmin=0, vmax=vmax)
+cb = plt.colorbar(art, ax=ax, label='z')
+ax.set_title('Fit')
+
+ax = axs[1, 0]
+fit = model.func(X, Y, **result.best_values)
+art = ax.pcolor(X, Y, Z-fit, vmin=0, vmax=10)
+cb = plt.colorbar(art, ax=ax, label='z')
+ax.set_title('Data - Fit')
+
+
+def set_label(ax):
+    ax.set_xlabel('x')
+    ax.set_ylabel('y')
+
+
+[set_label(ax) for ax in axs.ravel()]
+axs[1, 1].remove()
+
+###############################################################################
+# Two-dimensional off-axis Lorentzian
+# -----------------------------------
+# We now go on to show a harder example, in which the peak has a Lorentzian
+# profile and an off-axis anisotropic shape. This can be handled by applying a
+# suitable coordinate transform and then using the `lorentzian` function that
+# lmfit provides in the lineshapes module.
+def lorentzian2d(x, y, amplitude=1., centerx=0., centery=0., sigmax=1., sigmay=1.,
+                 rotation=0):
+    """Return a two dimensional lorentzian.
+
+    The maximum of the peak occurs at ``centerx`` and ``centery``
+    with widths ``sigmax`` and ``sigmay`` in the x and y directions
+    respectively. The peak can be rotated by choosing the value of ``rotation``
+    in radians.
+    """
+    xp = (x - centerx)*np.cos(rotation) - (y - centery)*np.sin(rotation)
+    yp = (x - centerx)*np.sin(rotation) + (y - centery)*np.cos(rotation)
+    R = (xp/sigmax)**2 + (yp/sigmay)**2
+
+    return 2*amplitude*lorentzian(R)/(np.pi*sigmax*sigmay)
+
```

Comment:
this is really nice and clearly written example for the gallery - thanks a lot!

Just out of curiosity (I haven't looked at the math or thought this thru, so it might be a stupid question): would it be possible to generalize this by making a CompositeModel from a built-in Guassian2D (or Lorentzian2D) model and some functional form to introduce the rotation?

Reply:
It's a very interesting question. 

I think, however, that one cannot build the function this way with the current construction of CompositeModel. Currently, composite model is constructed as a function of two arrays (which are made by evaluating the two functions). 

The most general way to combine two functions would be a functional i.e.

``def composite_function(func1, func2, x, params):``

My feeling is that it might be better to implement CompositeModel in this way, or to construct a new class that works in this way.